### ISSUE TITLE
Depending on jobs that may fail, but that's expected.

### ISSUE THREAD

**Issue by bloodearnest** [opensafely-core/job-runner#713]
The folks from Bristol have common pattern where they run a lot of parameterised models in parallel.  They expect some of these models will fail, as there will not be enough events, or they will fail to converge.  In their case this is actually fine, its part of what they are researching.

They have a final action that consumes all the outputs of models, but if any of them have failed, then this action won't run. So after they've run the parameterised model actions, they manually check actions succeeded, and modify/commit the project.yaml to just depend on the ones that succeeded, so they are able to submit and run that final action.

This manual step is laborious, especially as there can be a lot of actions, and visually inspecting the status can be time consuming.

## Potential solutions

They have requested a structured way to interrogate a workspace for the latest status of actions, so they can automate the modification of the dependencies in project.yaml (e.g. a downloadable json file or similar of actions and success/failure). Whilst this does make the manual step much easier, it doesn't remove the problem of having modify the final actions dependencies.

As a workaround, Will suggested they may be able to wrap their model code in error handling that will mean that it *always* succeeds, and use the contents of the output file as an indication of whether it materially succeed or not. This will mean they don't need to modify their project.yaml, but it does obfuscate whether a particular model succeeded or not, both in production and locally.

Another option would be to provide a way to mark an action as being allowed to fail, i.e. its failure will not block dependencies from running, but there will be no outputs from it.  Downstream actions can use the presence or absence of the expected output files as a signal to know if the model was successful or not.  This is a more general solution, that may have other use cases too.

**Comment by LFISHER7** [comment]
I agree with Dave that Will's suggestion is the right way to do things. These jobs failed because the script failed to handle errors that you might expect when fitting these models. Though I accept that handling these errors is a little tedious and deviates from current ways of working.

What we need is better [experimentation tracking](https://wandb.ai/site/experiment-tracking). This would require some rethinking of how outputs are captured in OpenSAFELY. [DVC experiment pipelines](https://dvc.org/doc/start/experiments/experiment-pipelines) seem like the kind of thing we'd want.

**Comment by evansd** [comment]
Ah, super useful to see prior art on this. Thanks Louis!
